ihmc-high-level-behaviors/src/main/java/us/ihmc/perception/RapidHeightMapDriftOffset.java

```diff
@@ -0,0 +1,61 @@
+package us.ihmc.perception;
+
+import controller_msgs.msg.dds.PlanOffsetStatus;
+import us.ihmc.behaviors.activeMapping.ControllerFootstepQueueMonitor;
+import us.ihmc.euclid.tuple3D.Vector3D;
+
+/**
+ * This class is meant to keep track of the drift in Z that is being calculated in the controller on the robot.
+ * This class takes in a {@link ControllerFootstepQueueMonitor} in order to subscribe to the {@link PlanOffsetStatus}.
+ * It then keeps track of the current drift since walking has started.
+ */
+public class RapidHeightMapDriftOffset
+{
+   private final ControllerFootstepQueueMonitor controllerFootstepQueueMonitor;
+   private final Vector3D mostRecentPlanOffsetProcessed = new Vector3D();
+   private final Vector3D incrementalOffset = new Vector3D();
+
+   public RapidHeightMapDriftOffset(ControllerFootstepQueueMonitor controllerFootstepQueueMonitor)
+   {
+      this.controllerFootstepQueueMonitor = controllerFootstepQueueMonitor;
+   }
+
+   public void reset()
+   {
+      mostRecentPlanOffsetProcessed.setToZero();
+   }
+
+   /**
+    * The method returns the amount of drift that has happened in meters from the controller.
+    * If no drift has happened then {@link Float#NaN} is returned
+    */
+   public float getUpdateDriftOffset()
+   {
+      incrementalOffset.setToNaN();
+
+      if (controllerFootstepQueueMonitor.isWalkingStarted())
+      {
+         // We reset this because the controller resets the drift on its end. So we need to reset ours as well.
+         // The existing drift is already captured in the height map by the previous offsets
+         mostRecentPlanOffsetProcessed.setToZero();
+      }
+
+      // While the robot isn't walking, this message will be null
+      PlanOffsetStatus latestPlanOffsetMessage = controllerFootstepQueueMonitor.getPlanOffsetMessage();
+      if (latestPlanOffsetMessage == null)
+         return Float.NaN;
+
+      Vector3D latestPlanOffset = latestPlanOffsetMessage.getOffsetVector();
+
+      if (controllerFootstepQueueMonitor.isFootstepStarted() && latestPlanOffset != null)
+      {
+         incrementalOffset.set(latestPlanOffset);
+         // We subtract the drift that has already been accounted for to only adjust by the new drift
+         incrementalOffset.sub(mostRecentPlanOffsetProcessed);
+         // Our internal tracker needs to add the drift because that is the new value being accounted for, so add it to the running total
+         mostRecentPlanOffsetProcessed.add(incrementalOffset);
```

Comment:
Something about the logic here seems fishy to me. It seems like this line can be replaced with this: 
```suggestion
         mostRecentPlanOffsetProcessed.set(latestPlanOffset);
```
And that makes me question setting the `mostRecentPlanOffsetProcessed` to zero above. Am I misunderstanding something?

Reply:
Ahhh ok ya its because I didn't explain what's happening on the controller. And its a lot of text to explain but I'll try.

Reply:
So when the robot starts walking. The drift tracker from the controller is reset to 0. Meaning anything down the chain needs to account for that. So when the robot starts walking. I reset to zero drift.
What the controller is sending is the total amount of drift. So locally we track what was the last drift we accounted for. And subtract that from the new total. To get the new drift that we should adjust for.
So first, set the incrementalDrift to be the latest total amount of drift.
Second, subtract the drift that was used last time. Will be zero if its the first time.
Third, not that we have our incremental change to account for. Add that to the current total drift tracker (this is the `mostRecentPlanOffsetProcessed`).
And the checks from the `ControllerFootstepQueueMonitor` are to deal with how and when the drift is sent from the controller.

Reply:
PS. I'm doing these comments from IntelliJ so sorry if the formatting sucks lol